File: experiments/adrianco/experiment-50-brazil-80b-uncapped/brazil/runs/agent=hermes-local_language=python_model=mlxlocal/mlx-community--Qwen3-Coder-Next-4bit_prompt=neutral_stack=m80/rep3/mcp_server.py
"""
Brazilian Soccer MCP Server
A Model Context Protocol server for Brazilian soccer data
"""

import asyncio
import os
import re
from datetime import datetime
from typing import Optional, List, Dict, Any
from pathlib import Path
import pandas as pd

# MCP imports
from mcp.server import Server
from mcp.server.stdio import stdio_server
from mcp.types import TextContent, ImageContent, Tool, Prompt
from mcp.shared.exceptions import McpError
from mcp.server.models import InitializationOptions
from mcp.server.lowlevel.server import NotificationOptions

# Local imports
from data_loader import SoccerDataLoader
from query_engine import (
    SoccerQueryEngine,
    TeamStats,
    MatchResult,
    PlayerInfo,
    CompetitionStandings,
    BigWin,
)

# Get data directory path
DATA_DIR = Path(__file__).parent / "data" / "kaggle"

# Initialize components
data_loader = SoccerDataLoader(DATA_DIR)
query_engine = SoccerQueryEngine(data_loader)

# Create MCP server
server = Server("brazilian-soccer-server")


def format_matches_response(matches: List[Dict], title: str = "Matches") -> str:
    """Format matches into a readable response"""
    if not matches:
        return f"No matches found for: {title}"
    
    lines = [f"{title} ({len(matches)} matches):", ""]
    for match in matches[:20]:  # Limit to 20 matches
        date = match.get('datetime', '')
        home = match.get('home_team', '')
        away = match.get('away_team', '')
        home_goal = match.get('home_goal', 0)
        away_goal = match.get('away_goal', 0)
        competition = match.get('competition', '')
        round_num = match.get('round', '')
        
        round_info = f" (Round {round_num})" if round_num else ""
        lines.append(f"- {date}: {home} {home_goal}-{away_goal} {away} [{competition}{round_info}]")
    
    if len(matches) > 20:
        lines.append(f"")
        lines.append(f"... and {len(matches) - 20} more matches")
    
    return "\n".join(lines)


def format_team_stats_response(stats: TeamStats, team: str) -> str:
    """Format team statistics into a readable response"""
    return f"""{team} Statistics:
- Matches: {stats.matches}
- Wins: {stats.wins}, Draws: {stats.draws}, Losses: {stats.losses}
- Goals For: {stats.goals_for}, Goals Against: {stats.goals_against}
- Goal Difference: {stats.goal_difference}
- Points: {stats.points}
- Win Rate: {stats.win_rate}%"""


def format_head_to_head_response(h2h: Dict, team1: str, team2: str) -> str:
    """Format head-to-head record into a readable response"""
    return f"""{team1} vs {team2} Head-to-Head:
- {team1} Wins: {h2h['team1_wins']}
- {team2} Wins: {h2h['team2_wins']}
- Draws: {h2h['draws']}
- Total Matches: {h2h['total_matches']}
"""


def format_player_response(player: PlayerInfo) -> str:
    """Format player information into a readable response"""
    return f"""{player.name}:
- ID: {player.id}
- Age: {player.age}
- Nationality: {player.nationality}
- Overall Rating: {player.overall}
- Potential: {player.potential}
- Club: {player.club}
- Position: {player.position}
- Jersey Number: {player.jersey_number}
- Height: {player.height}
- Weight: {player.weight}"""


def format_players_response(players: List[PlayerInfo], title: str) -> str:
    """Format list of players into a readable response"""
    if not players:
        return f"No players found for: {title}"
    
    lines = [f"{title} ({len(players)} players):", ""]
    for player in players[:20]:  # Limit to 20 players
        lines.append(f"- {player.name} ({player.position}) - {player.club} (Rating: {player.overall})")
    
    if len(players) > 20:
        lines.append(f"")
        lines.append(f"... and {len(players) - 20} more players")
    
    return "\n".join(lines)


def format_standings_response(standings: List[CompetitionStandings], competition: str, season: int) -> str:
    """Format competition standings into a readable response"""
    if not standings:
        return f"No standings found for {competition} {season}"
    
    lines = [f"{competition} {season} Standings:", ""]
    for standing in standings[:20]:  # Show top 20
        pos = standing.position
        team = standing.team
        pts = standing.points
        w = standing.wins
        d = standing.draws
        l = standing.losses
        gf = standing.goals_for
        ga = standing.goals_against
        gd = standing.goal_difference
        
        lines.append(f"{pos}. {team} - {pts} pts ({w}W, {d}D, {l}L) GF:{gf} GA:{ga} GD:{gd}")
    
    return "\n".join(lines)


def format_big_wins_response(wins: List[BigWin], limit: int) -> str:
    """Format big wins into a readable response"""
    if not wins:
        return f"No big wins found"
    
    lines = [f"Biggest Victories (Score difference >= 4 goals):", ""]
    for win in wins[:limit]:
        date = win.date
        home = win.home_team
        away = win.away_team
        home_goal = win.home_goal
        away_goal = win.away_goal
        competition = win.competition
        season = win.season
        
        season_info = f" ({season})" if season else ""
        lines.append(f"- {date}: {home} {home_goal}-{away_goal} {away} [{competition}{season_info}]")
    
    return "\n".join(lines)


# ========== Tool Handlers ==========

@server.list_tools()
async def list_tools() -> List[Tool]:
    """List all available tools for the MCP server"""
    return [
        Tool(
            name="find_matches_between_teams",
            description="Find all matches between two teams",
            inputSchema={
                "type": "object",
                "properties": {
                    "team1": {"type": "string", "description": "First team name"},
                    "team2": {"type": "string", "description": "Second team name"},
                },
                "required": ["team1", "team2"],
            },
        ),
        Tool(
            name="find_matches_by_team",
            description="Find matches for a specific team",
            inputSchema={
                "type": "object",
                "properties": {
                    "team": {"type": "string", "description": "Team name"},
                    "limit": {"type": "integer", "description": "Maximum number of matches to return", "default": 10},
                },
                "required": ["team"],
            },
        ),
        Tool(
            name="find_matches_by_competition",
            description="Find matches by competition type",
            inputSchema={
                "type": "object",
                "properties": {
                    "competition": {"type": "string", "description": "Competition name (Brasileirao, Copa do Brasil, Libertadores)"},
                    "season": {"type": "integer", "description": "Season year"},
                    "limit": {"type": "integer", "description": "Maximum number of matches to return", "default": 10},
                },
                "required": ["competition"],
            },
        ),
        Tool(
            name="find_matches_by_date_range",
            description="Find matches within a date range",
            inputSchema={
                "type": "object",
                "properties": {
                    "date_from": {"type": "string", "description": "Start date (YYYY-MM-DD)"},
                    "date_to": {"type": "string", "description": "End date (YYYY-MM-DD)"},
                    "team": {"type": "string", "description": "Optional team filter"},
                },
                "required": ["date_from", "date_to"],
            },
        ),
        Tool(
            name="get_team_statistics",
            description="Get comprehensive statistics for a team",
            inputSchema={
                "type": "object",
                "properties": {
                    "team": {"type": "string", "description": "Team name"},
                    "season": {"type": "integer", "description": "Season year (optional)"},
                    "competition": {"type": "string", "description": "Competition name (optional)"},
                },
                "required": ["team"],
            },
        ),
        Tool(
            name="get_head_to_head",
            description="Get head-to-head record between two teams",
            inputSchema={
                "type": "object",
                "properties": {
                    "team1": {"type": "string", "description": "First team name"},
                    "team2": {"type": "string", "description": "Second team name"},
                },
                "required": ["team1", "team2"],
            },
        ),
        Tool(
            name="get_player_by_name",
            description="Find a player by name",
            inputSchema={
                "type": "object",
                "properties": {
                    "name": {"type": "string", "description": "Player name to search"},
                },
                "required": ["name"],
            },
        ),
        Tool(
            name="get_players_by_club",
            description="Get all players from a club",
            inputSchema={
                "type": "object",
                "properties": {
                    "club": {"type": "string", "description": "Club name"},
                },
                "required": ["club"],
            },
        ),
        Tool(
            name="get_brazilian_players",
            description="Get all Brazilian players",
            inputSchema={
                "type": "object",
                "properties": {},
                "required": [],
            },
        ),
        Tool(
            name="get_competition_standings",
            description="Get competition standings for a season",
            inputSchema={
                "type": "object",
                "properties": {
                    "competition": {"type": "string", "description": "Competition name"},
                    "season": {"type": "integer", "description": "Season year"},
                },
                "required": ["competition", "season"],
            },
        ),
        Tool(
            name="get_big_wins",
            description="Get biggest wins in the dataset",
            inputSchema={
                "type": "object",
                "properties": {
                    "limit": {"type": "integer", "description": "Number of big wins to return", "default": 10},
                },
                "required": [],
            },
        ),
    ]


@server.call_tool()
async def call_tool(name: str, arguments: Dict[str, Any]) -> List[TextContent]:
    """Handle tool calls from the MCP client"""
    try:
        if name == "find_matches_between_teams":
            team1 = arguments.get("team1", "")
            team2 = arguments.get("team2", "")
            matches = query_engine.find_matches_between_teams(team1, team2)
            return [TextContent(text=format_matches_response(matches, f"{team1} vs {team2}"))]
        
        elif name == "find_matches_by_team":
            team = arguments.get("team", "")
            limit = arguments.get("limit", 10)
            matches = query_engine.find_matches_by_team(team, limit)
            return [TextContent(text=format_matches_response(matches, f"{team} Matches"))]
        
        elif name == "find_matches_by_competition":
            competition = arguments.get("competition", "")
            season = arguments.get("season")
            limit = arguments.get("limit", 10)
            
            # Get matches for the competition
            matches = query_engine.find_matches_by_team(competition, limit)
            if season:
                # Filter by season if provided
                matches = [m for m in matches if m.get('season') == season]
            
            return [TextContent(text=format_matches_response(matches, f"{competition} Matches"))]
        
        elif name == "find_matches_by_date_range":
            date_from = arguments.get("date_from", "")
            date_to = arguments.get("date_to", "")
            team = arguments.get("team", "")
            
            # Use get_all_matches with date filters
            all_matches = data_loader.get_all_matches(
                date_from=date_from,
                date_to=date_to,
                home_team=team if team else None,
                away_team=team if team else None,
            )
            
            return [TextContent(text=format_matches_response(all_matches, f"Matches {date_from} to {date_to}"))]
        
        elif name == "get_team_statistics":
            team = arguments.get("team", "")
            season = arguments.get("season")
            competition = arguments.get("competition")
            
            stats = query_engine.get_team_statistics(team, season, competition)
            return [TextContent(text=format_team_stats_response(stats, team))]
        
        elif name == "get_head_to_head":
            team1 = arguments.get("team1", "")
            team2 = arguments.get("team2", "")
            
            h2h = query_engine.get_head_to_head(team1, team2)
            return [TextContent(text=format_head_to_head_response(h2h, team1, team2))]
        
        elif name == "get_player_by_name":
            name = arguments.get("name", "")
            player = query_engine.get_player_by_name(name)
            
            if player:
                return [TextContent(text=format_player_response(player))]
            else:
                return [TextContent(text=f"Player '{name}' not found")]
        
        elif name == "get_players_by_club":
            club = arguments.get("club", "")
            players = query_engine.get_players_by_club(club)
            return [TextContent(text=format_players_response(players, f"{club} Players"))]
        
        elif name == "get_brazilian_players":
            players = query_engine.get_brazilian_players()
            return [TextContent(text=format_players_response(players, "Brazilian Players"))]
        
        elif name == "get_competition_standings":
            competition = arguments.get("competition", "")
            season = arguments.get("season", 2012)
            
            standings = query_engine.get_competition_standings(competition, season)
            return [TextContent(text=format_standings_response(standings, competition, season))]
        
        elif name == "get_big_wins":
            limit = arguments.get("limit", 10)
            wins = query_engine.get_big_wins(limit)
            return [TextContent(text=format_big_wins_response(wins, limit))]
        
        else:
            return [TextContent(text=f"Unknown tool: {name}")]
    
    except Exception as e:
        return [TextContent(text=f"Error: {str(e)}")]


# ========== Main Entry Point ==========

async def main():
    """Main entry point for the MCP server"""
    async with stdio_server() as (read_stream, write_stream):
        await server.run(
            read_stream,
            write_stream,
            InitializationOptions(
                server_name="brazilian-soccer-server",
                server_version="1.0.0",
                capabilities=server.get_capabilities(
                    notification_options=NotificationOptions(),
                    experimental_capabilities={},
                ),
            ),
        )


if __name__ == "__main__":
    asyncio.run(main())

File: experiments/adrianco/experiment-58-sol/brazil/runs/agent=codex_effort=default_language=typescript_model=gpt-5.6-sol_prompt=neutral/rep1-failed/src/mcp-server.ts
import type { Readable, Writable } from "node:stream";
import { createInterface } from "node:readline";
import type { SoccerService } from "./soccer-service.js";

interface JsonRpcRequest {
  jsonrpc: "2.0";
  id?: string | number | null;
  method: string;
  params?: Record<string, unknown>;
}

interface JsonRpcResponse {
  jsonrpc: "2.0";
  id: string | number | null;
  result?: unknown;
  error?: { code: number; message: string; data?: unknown };
}

type JsonSchema = Record<string, unknown>;

interface ToolDefinition {
  name: string;
  title: string;
  description: string;
  inputSchema: JsonSchema;
}

const objectSchema = (properties: Record<string, JsonSchema>, required: string[] = []): JsonSchema => ({
  type: "object",
  properties,
  ...(required.length > 0 ? { required } : {}),
  additionalProperties: false,
});

const string = (description: string, values?: string[]): JsonSchema => ({ type: "string", description, ...(values ? { enum: values } : {}) });
const integer = (description: string, minimum?: number, maximum?: number): JsonSchema => ({ type: "integer", description, ...(minimum !== undefined ? { minimum } : {}), ...(maximum !== undefined ? { maximum } : {}) });

export const TOOL_DEFINITIONS: ToolDefinition[] = [
  {
    name: "search_matches",
    title: "Search Brazilian Soccer Matches",
    description: "Search all five match datasets by team, opponent, venue, date range, competition, season, round, or stage. Team names are accent-, suffix-, and alias-normalized.",
    inputSchema: objectSchema({
      team: string("Team playing home or away"), opponent: string("Opponent when team is provided"),
      homeTeam: string("Home team only"), awayTeam: string("Away team only"),
      competition: string("Competition, such as Brasileirão, Copa do Brasil, or Libertadores"),
      season: integer("Season year", 1900, 2100), dateFrom: string("Inclusive ISO date YYYY-MM-DD"), dateTo: string("Inclusive ISO date YYYY-MM-DD"),
      stage: string("Tournament stage"), round: string("Round label"), sort: string("Date order", ["asc", "desc"]),
      offset: integer("Pagination offset", 0), limit: integer("Maximum results", 1, 500),
    }),
  },
  {
    name: "get_team_statistics",
    title: "Get Team Statistics",
    description: "Calculate a team's matches, wins, draws, losses, goals, goal difference, points, and win rate, optionally by season, competition, and venue.",
    inputSchema: objectSchema({ team: string("Team name"), season: integer("Season year", 1900, 2100), competition: string("Competition"), venue: string("Match venue", ["home", "away", "all"]) }, ["team"]),
  },
  {
    name: "compare_teams",
    title: "Compare Teams Head-to-Head",
    description: "Return head-to-head matches and each team's record against the other.",
    inputSchema: objectSchema({ team1: string("First team"), team2: string("Second team"), season: integer("Optional season", 1900, 2100), competition: string("Optional competition") }, ["team1", "team2"]),
  },
  {
    name: "search_players",
    title: "Search FIFA Players",
    description: "Search FIFA players by name, nationality, club, position or position family, rating, and age.",
    inputSchema: objectSchema({ name: string("Partial player name"), nationality: string("Nationality"), club: string("Club name"), position: string("Position code or forward/midfielder/defender/goalkeeper"), minOverall: integer("Minimum overall rating", 0, 100), maxAge: integer("Maximum age", 14, 60), limit: integer("Maximum results", 1, 500) }),
  },
  {
    name: "get_standings",
    title: "Calculate Competition Standings",
    description: "Calculate a season table from match results using 3 points for a win and standard tie-break ordering.",
    inputSchema: objectSchema({ season: integer("Season year", 1900, 2100), competition: string("Competition; defaults to Brasileirão Série A") }, ["season"]),
  },
  {
    name: "get_competition_statistics",
    title: "Analyze a Competition",
    description: "Calculate goals per match, home/away/draw rates, and biggest wins for a competition and optional season.",
    inputSchema: objectSchema({ competition: string("Competition"), season: integer("Optional season", 1900, 2100) }, ["competition"]),
  },
  {
    name: "explore_soccer_graph",
    title: "Explore the Soccer Knowledge Graph",
    description: "Return a bounded node/edge subgraph connecting teams, matches, competitions, and FIFA players.",
    inputSchema: objectSchema({ team: string("Center graph on a team"), player: string("Filter player nodes by name"), competition: string("Filter matches by competition"), season: integer("Filter matches by season", 1900, 2100), limit: integer("Maximum match and player results", 1, 100) }),
  },
  {
    name: "answer_soccer_question",
    title: "Answer a Natural-Language Soccer Question",
    description: "Route a natural-language question to match, team, player, competition, standings, derby, or statistical analysis.",
    inputSchema: objectSchema({ question: string("Question about the supplied Brazilian soccer datasets") }, ["question"]),
  },
  {
    name: "dataset_summary",
    title: "Describe Dataset Coverage",
    description: "Return row counts, date coverage, team and competition counts, and per-file loading evidence.",
    inputSchema: objectSchema({}),
  },
];

const SAMPLE_QUESTIONS = [
  "Show me all Flamengo vs Fluminense matches",
  "What matches did Palmeiras play in 2023?",
  "Find all Copa do Brasil finals",
  "What is Corinthians' home record in 2022?",
  "Which team scored the most goals in Serie A 2023?",
  "Compare Palmeiras and Santos head-to-head",
  "Find all Brazilian players in the dataset",
  "Who are the highest-rated players at Flamengo?",
  "Show me all forwards from São Paulo FC",
  "Who won the 2019 Brasileirão?",
  "Show the 2018 Copa Libertadores bracket",
  "Which teams were relegated in 2020?",
  "What's the average goals per match in the Brasileirão?",
  "Which team has the best away record in 2023?",
  "Show me the biggest wins in the dataset",
  "When did Flamengo last play Corinthians?",
  "Who is Gabriel Barbosa?",
  "Which players play for Flamengo?",
  "Show me all derbies in 2023",
  "What competitions has Palmeiras played in?",
];

function toolResult(value: unknown): Record<string, unknown> {
  return {
    content: [{ type: "text", text: typeof value === "string" ? value : JSON.stringify(value, null, 2) }],
    structuredContent: value,
  };
}

function requiredString(args: Record<string, unknown>, name: string): string {
  const value = args[name];
  if (typeof value !== "string" || value.trim() === "") throw new Error(`'${name}' must be a non-empty string`);
  return value;
}

function optionalString(args: Record<string, unknown>, name: string): string | undefined {
  return typeof args[name] === "string" ? args[name] : undefined;
}

function optionalNumber(args: Record<string, unknown>, name: string): number | undefined {
  return typeof args[name] === "number" && Number.isFinite(args[name]) ? args[name] : undefined;
}

export class BrazilianSoccerMcpServer {
  constructor(private readonly service: SoccerService) {}

  async handle(request: JsonRpcRequest): Promise<JsonRpcResponse | null> {
    if (request.id === undefined) return null;
    const id = request.id ?? null;
    try {
      switch (request.method) {
        case "initialize": {
          const requested = typeof request.params?.protocolVersion === "string" ? request.params.protocolVersion : "2025-11-25";
          return this.ok(id, {
            protocolVersion: requested,
            capabilities: { tools: { listChanged: false }, resources: { subscribe: false, listChanged: false }, prompts: { listChanged: false } },
            serverInfo: { name: "brazilian-soccer-mcp", version: "1.0.0" },
            instructions: "Use answer_soccer_question for natural language, or the typed analytics tools for precise filters. All data is read-only and local; scorer events are not present.",
          });
        }
        case "ping": return this.ok(id, {});
        case "tools/list": return this.ok(id, { tools: TOOL_DEFINITIONS.map((tool) => ({ ...tool, annotations: { readOnlyHint: true, destructiveHint: false, idempotentHint: true, openWorldHint: false } })) });
        case "tools/call": return this.ok(id, await this.callTool(request.params ?? {}));
        case "resources/list": return this.ok(id, { resources: [
          { uri: "soccer://dataset/summary", name: "Brazilian Soccer Dataset Summary", mimeType: "application/json", description: "Coverage and per-source row counts" },
          { uri: "soccer://dataset/questions", name: "Example Soccer Questions", mimeType: "application/json", description: "Twenty supported question patterns" },
        ] });
        case "resources/read": return this.ok(id, this.readResource(requiredString(request.params ?? {}, "uri")));
        case "prompts/list": return this.ok(id, { prompts: [{ name: "analyze_brazilian_soccer", title: "Analyze Brazilian Soccer", description: "Ask a grounded question about the bundled datasets", arguments: [{ name: "question", description: "Your soccer question", required: true }] }] });
        case "prompts/get": return this.ok(id, this.getPrompt(request.params ?? {}));
        default: return { jsonrpc: "2.0", id, error: { code: -32601, message: `Method not found: ${request.method}` } };
      }
    } catch (error) {
      return { jsonrpc: "2.0", id, error: { code: -32602, message: error instanceof Error ? error.message : String(error) } };
    }
  }

  private ok(id: string | number | null, result: unknown): JsonRpcResponse {
    return { jsonrpc: "2.0", id, result };
  }

  private async callTool(params: Record<string, unknown>): Promise<Record<string, unknown>> {
    const name = requiredString(params, "name");
    const args = typeof params.arguments === "object" && params.arguments !== null ? params.arguments as Record<string, unknown> : {};
    switch (name) {
      case "search_matches": return toolResult(this.service.searchMatches({
        ...this.matchFilters(args),
        ...(optionalString(args, "homeTeam") ? { homeTeam: optionalString(args, "homeTeam") } : {}),
        ...(optionalString(args, "awayTeam") ? { awayTeam: optionalString(args, "awayTeam") } : {}),
        ...(optionalString(args, "dateFrom") ? { dateFrom: optionalString(args, "dateFrom") } : {}),
        ...(optionalString(args, "dateTo") ? { dateTo: optionalString(args, "dateTo") } : {}),
        ...(optionalString(args, "stage") ? { stage: optionalString(args, "stage") } : {}),
        ...(optionalString(args, "round") ? { round: optionalString(args, "round") } : {}),
        ...(optionalString(args, "sort") === "asc" || optionalString(args, "sort") === "desc" ? { sort: optionalString(args, "sort") as "asc" | "desc" } : {}),
        ...(optionalNumber(args, "offset") !== undefined ? { offset: optionalNumber(args, "offset") } : {}),
        ...(optionalNumber(args, "limit") !== undefined ? { limit: optionalNumber(args, "limit") } : {}),
      }));
      case "get_team_statistics": return toolResult(this.service.getTeamStatistics(requiredString(args, "team"), {
        ...this.statFilters(args),
        ...(optionalString(args, "venue") ? { venue: optionalString(args, "venue") as "home" | "away" | "all" } : {}),
      }));
      case "compare_teams": return toolResult(this.service.compareTeams(requiredString(args, "team1"), requiredString(args, "team2"), this.statFilters(args)));
      case "search_players": return toolResult(this.service.searchPlayers({
        ...(optionalString(args, "name") ? { name: optionalString(args, "name") } : {}),
        ...(optionalString(args, "nationality") ? { nationality: optionalString(args, "nationality") } : {}),
        ...(optionalString(args, "club") ? { club: optionalString(args, "club") } : {}),
        ...(optionalString(args, "position") ? { position: optionalString(args, "position") } : {}),
        ...(optionalNumber(args, "minOverall") !== undefined ? { minOverall: optionalNumber(args, "minOverall") } : {}),
        ...(optionalNumber(args, "maxAge") !== undefined ? { maxAge: optionalNumber(args, "maxAge") } : {}),
        ...(optionalNumber(args, "limit") !== undefined ? { limit: optionalNumber(args, "limit") } : {}),
      }));
      case "get_standings": return toolResult(this.service.getStandings(optionalNumber(args, "season") ?? (() => { throw new Error("'season' must be a number"); })(), optionalString(args, "competition")));
      case "get_competition_statistics": return toolResult(this.service.getCompetitionStatistics(requiredString(args, "competition"), optionalNumber(args, "season")));
      case "explore_soccer_graph": return toolResult(this.service.exploreGraph({
        ...(optionalString(args, "team") ? { team: optionalString(args, "team") } : {}),
        ...(optionalString(args, "player") ? { player: optionalString(args, "player") } : {}),
        ...(optionalString(args, "competition") ? { competition: optionalString(args, "competition") } : {}),
        ...(optionalNumber(args, "season") !== undefined ? { season: optionalNumber(args, "season") } : {}),
        ...(optionalNumber(args, "limit") !== undefined ? { limit: optionalNumber(args, "limit") } : {}),
      }));
      case "answer_soccer_question": return toolResult(this.service.answerQuestion(requiredString(args, "question")));
      case "dataset_summary": return toolResult(this.service.data.summary);
      default: throw new Error(`Unknown tool: ${name}`);
    }
  }

  private matchFilters(args: Record<string, unknown>): Record<string, unknown> {
    return {
      ...(optionalString(args, "team") ? { team: optionalString(args, "team") } : {}),
      ...(optionalString(args, "opponent") ? { opponent: optionalString(args, "opponent") } : {}),
      ...this.statFilters(args),
    };
  }

  private statFilters(args: Record<string, unknown>): Record<string, unknown> {
    return {
      ...(optionalNumber(args, "season") !== undefined ? { season: optionalNumber(args, "season") } : {}),
      ...(optionalString(args, "competition") ? { competition: optionalString(args, "competition") } : {}),
    };
  }

  private readResource(uri: string): { contents: Array<Record<string, unknown>> } {
    if (uri === "soccer://dataset/summary") return { contents: [{ uri, mimeType: "application/json", text: JSON.stringify(this.service.data.summary, null, 2) }] };
    if (uri === "soccer://dataset/questions") return { contents: [{ uri, mimeType: "application/json", text: JSON.stringify(SAMPLE_QUESTIONS, null, 2) }] };
    throw new Error(`Unknown resource: ${uri}`);
  }

  private getPrompt(params: Record<string, unknown>): Record<string, unknown> {
    if (requiredString(params, "name") !== "analyze_brazilian_soccer") throw new Error("Unknown prompt");
    const args = typeof params.arguments === "object" && params.arguments !== null ? params.arguments as Record<string, unknown> : {};
    const question = requiredString(args, "question");
    return { description: "Analyze a question using only the bundled soccer datasets", messages: [{ role: "user", content: { type: "text", text: `Use the Brazilian soccer tools to answer this question. State when a requested fact is not inferable from the data.\n\n${question}` } }] };
  }
}

export async function serveStdio(server: BrazilianSoccerMcpServer, input: Readable = process.stdin, output: Writable = process.stdout): Promise<void> {
  const lines = createInterface({ input, crlfDelay: Infinity });
  for await (const line of lines) {
    if (!line.trim()) continue;
    let response: JsonRpcResponse | null;
    try {
      response = await server.handle(JSON.parse(line) as JsonRpcRequest);
    } catch (error) {
      response = { jsonrpc: "2.0", id: null, error: { code: -32700, message: "Parse error", data: error instanceof Error ? error.message : String(error) } };
    }
    if (response) output.write(`${JSON.stringify(response)}\n`);
  }
}
